### DISCUSSION TITLE
Mobile auth best practices

### DISCUSSION THREAD

**Discussion by nugmanoff** [MobileNativeFoundation/discussions#75] [General]
Hello everyone!

I am curious to know what are some of the community best practices for handling Authentication of mobile apps.
There is [RFC 8252 - OAuth 2.0 for Native Apps](https://tools.ietf.org/html/rfc8252) – and it describes the approach to authentication using web-agents (namely `SFSafariViewController` on iOS) for this purposes. 

I understand the security implications, but there are implications in terms of user experience as well. Natively built interface for authentication is much more pleasant in terms of experience rather than embedded web browser.

For those who follow conventional OAuth 2.0 protocol with their backend:

1. What is the lifetime of `access_token` and `refresh_token`?
2. How do you handle `refresh_token` expiration – immediate logout or you store the user password in encrypted storage (keychain) and then authorize the user behind the scenes?

There are many apps that leverage the usage of PIN + Biometric Auth in order to reduce the amount of friction that is needed to authorize user (typing whole password vs entering short pin, scanning faceid/touchid). Yet the question is the same in this case, how is it implemented behind the scenes? (long-lived tokens? pin is send to backend to validate?)

All in all, the question is how to provide frictionless user experience for authorization at the same time with adhering to security guidelines.

**Comment by Exadra37** [comment]
> pin is send to backend to validate?

OAUTH validations, issue of tokens, and tokens refresh must always be done in the backend, otherwise if they are done in the client side, then they can be bypassed easily with the use of instrumentation frameworks at runtime, like by using https://frida.re.

Regarding your other questions I will let for the mobile developers to answer them, because I am just a Developer Advocate for security in mobile APIs.
